File: bridge-pattern/js/src/RefinedShape.js
import { AbstractShape } from './AbstractShape.js'

// 具体桥接类，继承抽象桥接类，扩充了自己的函数
// 通过依赖的工具来辅助具体功能
export class RefinedShape extends AbstractShape {
  constructor(drawTool) {
    super(drawTool)
  }

  draw(x, y, radius) {
    console.log(`RefinedShape::draw() [x=${x} y=${y} radius=${radius}]`)
    // 通过工具类实际绘制
    this.drawTool.drawStyle()
  }

  // 新抽象类里的方法，可作为补充
  drawDone() {
    console.log('RefinedShape::drawDone(), 执行的drawTool是: ' + this.drawTool.constructor.name)
  }
}

File: bridge-pattern/go/src/RefinedShape.go
package src

import (
  "fmt"
  "strconv"
)

// 具体桥接类，继承抽象桥接类，扩充了自己的方法
// 也可以将此类当做抽象类，再往下派生具体实现类
type RefinedShape struct {
  Name     string
  drawTool DrawTool
}

// 设置具体工具类
func (c *RefinedShape) SetDrawTool(drawTool DrawTool) {
  fmt.Println("RefinedShape::SetDrawTool() [drawTool.Name=" + drawTool.GetName() + "]")
  c.drawTool = drawTool
}

func (c *RefinedShape) GetDrawTool() DrawTool {
  return c.drawTool
}

func (c *RefinedShape) Draw(x int, y int, radius int) {
  fmt.Println("RefinedShape::Draw() [x=" + strconv.Itoa(x) +
    " y=" + strconv.Itoa(y) + " radius=" + strconv.Itoa(radius) + "]")
  // 通过工具类实现具体功能
  if c.drawTool != nil {
    c.drawTool.DrawStyle()
  }
}

// 新方法，可作为具体工具的补充
func (r *RefinedShape) DrawDone(shape AbstractShape) {
  fmt.Println("RefinedShape::DrawDone(), 执行的DrawTool是: ", shape.GetDrawTool())
}
